File: Sources/DopamineCore/NLP.swift
import Foundation

/// Lightweight tokenization/vectorization utilities for project-topic inference.
public enum NLP {
    private static let stopWords: Set<String> = [
        "the", "a", "an", "and", "or", "to", "of", "in", "for", "on", "with", "is", "it", "this",
        "that", "i", "we", "you", "my", "our", "be", "at", "as", "by", "from", "me"
    ]

    public static func tokenize(_ text: String) -> [String] {
        let lower = text.lowercased()
        let clean = lower.replacingOccurrences(of: "[^a-z0-9\\s]", with: " ", options: .regularExpression)
        return clean
            .split(whereSeparator: { $0.isWhitespace })
            .map(String.init)
            .filter { $0.count > 1 && !stopWords.contains($0) }
    }

    public static func vectorize(_ text: String) -> [String: Double] {
        var counts: [String: Double] = [:]
        for token in tokenize(text) {
            counts[token, default: 0] += 1
        }

        let norm = sqrt(counts.values.reduce(0) { $0 + $1 * $1 })
        guard norm > 0 else { return counts }

        var normalized: [String: Double] = [:]
        for (token, value) in counts {
            normalized[token] = value / norm
        }
        return normalized
    }

    public static func cosineSimilarity(_ lhs: [String: Double], _ rhs: [String: Double]) -> Double {
        lhs.reduce(0) { partial, pair in
            partial + pair.value * (rhs[pair.key] ?? 0)
        }
    }

    public static func blendCentroid(
        centroid: [String: Double],
        vector: [String: Double],
        messageCount: Int
    ) -> [String: Double] {
        guard messageCount > 0 else { return vector }

        var next = centroid
        for (token, value) in vector {
            let previous = centroid[token] ?? 0
            next[token] = (previous * Double(messageCount) + value) / Double(messageCount + 1)
        }
        return next
    }
}

File: lib/nlp.ts
const STOP_WORDS = new Set([
  "the",
  "a",
  "an",
  "and",
  "or",
  "to",
  "of",
  "in",
  "for",
  "on",
  "with",
  "is",
  "it",
  "this",
  "that",
  "i",
  "we",
  "you",
  "my",
  "our",
  "be",
  "at",
  "as",
  "by",
  "from",
  "me"
]);

export function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, " ")
    .split(/\s+/)
    .filter((token) => token.length > 1 && !STOP_WORDS.has(token));
}

export function vectorize(text: string): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const token of tokenize(text)) {
    counts[token] = (counts[token] ?? 0) + 1;
  }
  const norm = Math.sqrt(Object.values(counts).reduce((sum, value) => sum + value * value, 0));
  if (!norm) {
    return counts;
  }
  const normalized: Record<string, number> = {};
  for (const [token, value] of Object.entries(counts)) {
    normalized[token] = value / norm;
  }
  return normalized;
}

export function cosineSimilarity(a: Record<string, number>, b: Record<string, number>): number {
  let score = 0;
  for (const [token, aValue] of Object.entries(a)) {
    const bValue = b[token] ?? 0;
    score += aValue * bValue;
  }
  return score;
}

export function blendCentroid(
  centroid: Record<string, number>,
  vector: Record<string, number>,
  messageCount: number
): Record<string, number> {
  if (messageCount <= 0) {
    return { ...vector };
  }
  const next: Record<string, number> = { ...centroid };
  for (const [token, value] of Object.entries(vector)) {
    next[token] = ((centroid[token] ?? 0) * messageCount + value) / (messageCount + 1);
  }
  return next;
}
